File: web/src/api/cards.ts
import { requestJson } from "./client";
import type {
  CreateMindCardRequest,
  CreateMindCardResponse,
  DeleteMindCardResponse,
  MyMindCardsResponse,
  PublicCardsResponse,
  SelectCardRequest,
  SelectCardResponse,
  UpdateMindCardRequest,
  UpdateMindCardResponse
} from "../types/card";

export function createMindCard(
  sessionId: string,
  request: CreateMindCardRequest
): Promise<CreateMindCardResponse> {
  return requestJson<CreateMindCardResponse>(`/sessions/${encodeURIComponent(sessionId)}/cards`, {
    method: "POST",
    body: request
  });
}

export function listMyMindCards(sessionId: string): Promise<MyMindCardsResponse> {
  return requestJson<MyMindCardsResponse>(`/sessions/${encodeURIComponent(sessionId)}/cards`);
}

export function getMyCards(sessionId: string): Promise<MyMindCardsResponse> {
  return listMyMindCards(sessionId);
}

export function updateMindCard(
  sessionId: string,
  cardId: string,
  request: UpdateMindCardRequest
): Promise<UpdateMindCardResponse> {
  return requestJson<UpdateMindCardResponse>(
    `/sessions/${encodeURIComponent(sessionId)}/cards/${encodeURIComponent(cardId)}`,
    {
      method: "PATCH",
      body: request
    }
  );
}

export function deleteMindCard(sessionId: string, cardId: string): Promise<DeleteMindCardResponse> {
  return requestJson<DeleteMindCardResponse>(
    `/sessions/${encodeURIComponent(sessionId)}/cards/${encodeURIComponent(cardId)}`,
    {
      method: "DELETE"
    }
  );
}

export function getPublicCards(
  eventSlug: string,
  excludeSessionId: string | undefined,
  limit = 10
): Promise<PublicCardsResponse> {
  const params = new URLSearchParams();
  params.set("limit", String(limit));
  if (excludeSessionId) {
    params.set("excludeSessionId", excludeSessionId);
  }
  return requestJson<PublicCardsResponse>(
    `/events/${encodeURIComponent(eventSlug)}/cards/public?${params.toString()}`
  );
}

export function selectCard(
  sessionId: string,
  request: SelectCardRequest
): Promise<SelectCardResponse> {
  return requestJson<SelectCardResponse>(`/sessions/${encodeURIComponent(sessionId)}/selected-card`, {
    method: "POST",
    body: request
  });
}

File: api/app/api/routes/cards.py
from uuid import UUID

from fastapi import APIRouter, Depends, Path, Query
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.cards import (
    CreateMindCardRequest,
    CreateMindCardResponse,
    DeleteMindCardResponse,
    MyMindCardsResponse,
    PublicMindCardsResponse,
    SelectCardRequest,
    SelectCardResponse,
    UpdateMindCardRequest,
    UpdateMindCardResponse,
)
from app.services.cards import (
    create_mind_card,
    delete_mind_card,
    list_my_cards,
    list_public_cards,
    select_peer_card,
    update_mind_card,
)

router = APIRouter()


@router.post("/sessions/{sessionId}/cards", response_model=CreateMindCardResponse)
def create_session_card(
    payload: CreateMindCardRequest,
    session_id: UUID = Path(alias="sessionId"),
    db: Session = Depends(get_db),
) -> CreateMindCardResponse:
    return create_mind_card(db, session_id, payload)


@router.get("/sessions/{sessionId}/cards", response_model=MyMindCardsResponse)
def read_session_cards(
    session_id: UUID = Path(alias="sessionId"),
    db: Session = Depends(get_db),
) -> MyMindCardsResponse:
    return list_my_cards(db, session_id)


@router.patch("/sessions/{sessionId}/cards/{cardId}", response_model=UpdateMindCardResponse)
def update_session_card(
    payload: UpdateMindCardRequest,
    session_id: UUID = Path(alias="sessionId"),
    card_id: UUID = Path(alias="cardId"),
    db: Session = Depends(get_db),
) -> UpdateMindCardResponse:
    return update_mind_card(db, session_id, card_id, payload)


@router.delete("/sessions/{sessionId}/cards/{cardId}", response_model=DeleteMindCardResponse)
def delete_session_card(
    session_id: UUID = Path(alias="sessionId"),
    card_id: UUID = Path(alias="cardId"),
    db: Session = Depends(get_db),
) -> DeleteMindCardResponse:
    return delete_mind_card(db, session_id, card_id)


@router.get("/events/{eventSlug}/cards/public", response_model=PublicMindCardsResponse)
def read_public_cards(
    event_slug: str = Path(alias="eventSlug"),
    exclude_session_id: UUID | None = Query(default=None, alias="excludeSessionId"),
    limit: int = Query(default=10, ge=1, le=30),
    db: Session = Depends(get_db),
) -> PublicMindCardsResponse:
    return list_public_cards(
        db,
        event_slug=event_slug,
        exclude_session_id=exclude_session_id,
        limit=limit,
    )


@router.post("/sessions/{sessionId}/selected-card", response_model=SelectCardResponse)
def select_session_card(
    payload: SelectCardRequest,
    session_id: UUID = Path(alias="sessionId"),
    db: Session = Depends(get_db),
) -> SelectCardResponse:
    return select_peer_card(db, session_id, payload)
